README.md
# payzen_dev

payment_payzen/CHANGELOG.md
1.2.1, 2020-01-21
=============
- Compatibility with Odoo 14 version.

1.2.0, 2020-04-07
=============
- Compatibility with Odoo 13 version.
- [multi] Added payment in installments submodule.

1.1.1, 2019-12-16
=============
- Bug fix: incorrect payment amount rounding.

1.1.0, 2019-06-17
=============
- Compatibility with Odoo 11 and 12 versions.
- Added Spanish translation.
- Enable signature algorithm selection (SHA-1 or HMAC-SHA-256).
- Added various configuration options.
- Improve payment details dislay.
- [technical] Manage enabled/disabled features by plugin variant.

1.0.0, 2017-05-01
=============
- Bug fix: authentication error on payment end when request contains accented characters.
- Rename some module files to match Odoo standards.
- Added shipping information in sent form.
- Consider AUTHORISED_TO_VALIDATE transaction status as a pending status.
- Added PayZen supported currencies.
- Added license notices.

0.9.1, 2017-05-01
=============
- Bug fix: relative to generating transaction IDs.
- Process return to shop and IPN calls.

0.9.0, 2017-01-22
=============
- Initial PayZen payment module for Odoo 9 by Sudokeys.

payment_payzen/__manifest__.py
# coding: utf-8
#
# Copyright © Lyra Network.
# This file is part of PayZen plugin for Odoo. See COPYING.md for license details.
#
# Author:    Lyra Network (https://www.lyra.com)
# Copyright: Copyright © Lyra Network
# License:   http://www.gnu.org/licenses/agpl.html GNU Affero General Public License (AGPL v3)

{
    'name': 'PayZen Payment Acquirer',
    'version': '1.2.1',
    'summary': 'Accept payments with PayZen secure payment gateway.',
    'category': 'Accounting',
    'author': 'Lyra Network',
    'website': 'https://www.lyra.com/',
    'license': 'AGPL-3',
    'depends': ['payment'],
    'data': [
        'views/payment_views.xml',
        'views/payment_payzen_templates.xml',
        'data/payment_acquirer_data.xml',
        'security/ir.model.access.csv',
    ],
    'images': ['static/description/icon.png'],
    'installable': True,
}


payment_payzen/helpers/constants.py
# coding: utf-8
#
# Copyright © Lyra Network.
# This file is part of PayZen plugin for Odoo. See COPYING.md for license details.
#
# Author:    Lyra Network (https://www.lyra.com)
# Copyright: Copyright © Lyra Network
# License:   http://www.gnu.org/licenses/agpl.html GNU Affero General Public License (AGPL v3)

from odoo import _

# WARN: Do not modify code format here. This is managed by build files.
PAYZEN_PLUGIN_FEATURES = {
    'multi': True,
    'restrictmulti': False,
    'qualif': False,
    'shatwo': True,
}

PAYZEN_PARAMS = {
    'GATEWAY_CODE': 'PayZen',
    'GATEWAY_NAME': 'PayZen',
    'BACKOFFICE_NAME': 'PayZen',
    'SUPPORT_EMAIL': 'support@payzen.eu',
    'GATEWAY_URL': 'https://secure.payzen.eu/vads-payment/',
    'SITE_ID': '12345678',
    'KEY_TEST': '1111111111111111',
    'KEY_PROD': '2222222222222222',
    'CTX_MODE': 'TEST',
    'SIGN_ALGO': 'SHA-256',
    'LANGUAGE': 'fr',

    'GATEWAY_VERSION': 'V2',
    'PLUGIN_VERSION': '1.2.1',
    'CMS_IDENTIFIER': 'Odoo_10-14',
}

PAYZEN_LANGUAGES = {
    'cn': 'Chinese',
    'de': 'German',
    'es': 'Spanish',
    'en': 'English',
    'fr': 'French',
    'it': 'Italian',
    'jp': 'Japanese',
    'nl': 'Dutch',
    'pl': 'Polish',
    'pt': 'Portuguese',
    'ru': 'Russian',
    'sv': 'Swedish',
    'tr': 'Turkish',
}

PAYZEN_CARDS = {
    'CB': u'CB',
    'E-CARTEBLEUE': u'e-Carte Bleue',
    'MAESTRO': u'Maestro',
    'MASTERCARD': u'Mastercard',
    'VISA': u'Visa',
    'VISA_ELECTRON': u'Visa Electron',
    'VPAY': u'V PAY',
    'AMEX': u'American Express',
    'ACCORD_STORE': u'Cartes Enseignes Partenaires',
    'ACCORD_STORE_SB': u'Cartes Enseignes Partenaires (sandbox)',
    'ALINEA_CDX': u'Carte Cadeau Alinéa',
    'ALINEA_CDX_SB': u'Carte Cadeau Alinéa (sandbox)',
    'ALIPAY': u'Alipay',
    'ALLOBEBE_CDX': u'Carte Cadeau Allobébé',
    'ALLOBEBE_CDX_SB': u'Carte Cadeau Allobébé (sandbox)',
    'APETIZ': u'Apetiz',
    'AUCHAN': u'Carte Auchan',
    'AUCHAN_SB': u'Carte Auchan (sandbox)',
    'AURORE-MULTI': u'Cpay Aurore',
    'BANCONTACT': u'Bancontact Mistercash',
    'BIZZBEE_CDX': u'Carte Cadeau Bizzbee',
    'BIZZBEE_CDX_SB': u'Carte Cadeau Bizzbee (sandbox)',
    'BOULANGER': u'Carte b+',
    'BOULANGER_SB': u'Carte b+ (sandbox)',
    'BRICE_CDX': u'Carte Cadeau Brice',
    'BRICE_CDX_SB': u'Carte Cadeau Brice (sandbox)',
    'BUT': u'But',
    'CABAL': u'Cabal',
    'CARNET': u'Carnet',
    'CA_DO_CARTE': u'CA DO Carte',
    'CHQ_DEJ': u'Chèque Déjeuner',
    'CONECS': u'Conecs',
    'CONFORAMA': u'Conforama',
    'CORA': u'Cora',
    'CORA_BLANCHE': u'Cora blanche',
    'CORA_PREM': u'Cora Visa Premier',
    'CORA_VISA': u'Cora Visa',
    'CVCO': u'Chèque-Vacances Connect',
    'DINERS': u'Diners',
    'DISCOVER': u'Discover',
    'ECCARD': u'EC Card',
    'EDENRED': u'Ticket Restaurant',
    'EDENRED_EC': u'Ticket EcoCheque',
    'EDENRED_SC': u'Ticket Sport & Culture',
    'EDENRED_TC': u'Ticket Compliments',
    'EDENRED_TR': u'Ticket Restaurant',
    'ELO': u'Elo',
    'E_CV': u'e-Chèque-Vacances',
    'FRANFINANCE_3X': u'Paiement en 3 fois',
    'FRANFINANCE_4X': u'Paiement en 4 fois',
    'FULLCB3X': u'Paiement en 3 fois CB',
    'FULLCB4X': u'Paiement en 4 fois CB',
    'GEMO_CDX': u'Carte Cadeau Gémo',
    'GEMO_CDX_SB': u'Carte Cadeau Gémo (sandbox)',
    'GIROPAY': u'Giropay',
    'GOOGLEPAY': u'Google Pay',
    'HIPER': u'Hiper',
    'HIPERCARD': u'Hipercard',
    'IDEAL': u'iDEAL',
    'JCB': u'JCB',
    'JOUECLUB_CDX': u'Carte Cadeau Joué Club',
    'JOUECLUB_CDX_SB': u'Carte Cadeau Joué Club (sandbox)',
    'JULES_CDX': u'Carte Cadeau Jules',
    'JULES_CDX_SB': u'Carte Cadeau Jules (sandbox)',
    'KLARNA': u'Klarna',
    'LECLERC': u'Carte Reglo',
    'MC_CORDOBESA': u'Mastercard Cordobesa',
    'MULTIBANCO': u'Multibanco',
    'MYBANK': u'MyBank',
    'NARANJA': u'Naranja',
    'NORAUTO': u'Carte Norauto option Financement',
    'NORAUTO_SB': u'Carte Norauto option Financement (sandbox)',
    'ONEY_3X_4X': u'Paiement en 3 ou 4 fois Oney',
    'ONEY_ENSEIGNE': u'Cartes enseignes Oney',
    'PAYDIREKT': u'Paydirekt',
    'PAYLIB': u'Paylib',
    'PAYPAL': u'PayPal',
    'PAYPAL_SB': u'PayPal Sandbox',
    'PICWIC': u'Carte Picwic',
    'PICWIC_SB': u'Carte Picwic (sandbox)',
    'POSTFINANCE': u'PostFinance Card',
    'POSTFINANCE_EFIN': u'PostFinance E-Finance',
    'PRESTO': u'Presto',
    'PRZELEWY24': u'Przelewy24',
    'S-MONEY': u'S-money',
    'SCT': u'Virement SEPA',
    'SDD': u'Prélèvement SEPA',
    'SODEXO': u'Pass Restaurant',
    'SOFORT_BANKING': u'Sofort',
    'SOROCRED': u'Sorocred',
    'TRUFFAUT_CDX': u'Carte Cadeau Truffaut',
    'UNION_PAY': u'UnionPay',
    'WECHAT': u'WeChat Pay',
}

PAYZEN_CURRENCIES = [
    ['AUD', '036', 2],
    ['KHR', '116', 0],
    ['CAD', '124', 2],
    ['CNY', '156', 1],
    ['CZK', '203', 2],
    ['DKK', '208', 2],
    ['HKD', '344', 2],
    ['HUF', '348', 2],
    ['INR', '356', 2],
    ['IDR', '360', 2],
    ['JPY', '392', 0],
    ['KRW', '410', 0],
    ['KWD', '414', 3],
    ['MYR', '458', 2],
    ['MXN', '484', 2],
    ['MAD', '504', 2],
    ['NZD', '554', 2],
    ['NOK', '578', 2],
    ['PHP', '608', 2],
    ['RUB', '643', 2],
    ['SGD', '702', 2],
    ['ZAR', '710', 2],
    ['SEK', '752', 2],
    ['CHF', '756', 2],
    ['THB', '764', 2],
    ['TND', '788', 3],
    ['GBP', '826', 2],
    ['USD', '840', 2],
    ['TWD', '901', 2],
    ['TRY', '949', 2],
    ['EUR', '978', 2],
    ['PLN', '985', 2],
    ['BRL', '986', 2],
    ['ARS', '032', 2],
    ['COP', '170', 2],
    ['PEN', '604', 2],
]


payment_payzen/models/payment.py
# coding: utf-8
#
# Copyright © Lyra Network.
# This file is part of PayZen plugin for Odoo. See COPYING.md for license details.
#
# Author:    Lyra Network (https://www.lyra.com)
# Copyright: Copyright © Lyra Network
# License:   http://www.gnu.org/licenses/agpl.html GNU Affero General Public License (AGPL v3)

import base64
from datetime import datetime
from hashlib import sha1, sha256
import hmac
import logging
import math
from os import path

from pkg_resources import parse_version

from odoo import models, api, release, fields, _
from odoo.addons.payment.models.payment_acquirer import ValidationError
from odoo.tools import convert_xml_import
from odoo.tools import float_round
from odoo.tools.float_utils import float_compare

from ..controllers.main import PayzenController
from ..helpers import constants, tools
from .card import PayzenCard
from .language import PayzenLanguage


try:
    import urlparse
except ImportError:
    import urllib.parse as urlparse

_logger = logging.getLogger(__name__)

class AcquirerPayzen(models.Model):
    _inherit = 'payment.acquirer'

    def _get_notify_url(self):
        base_url = self.env['ir.config_parameter'].get_param('web.base.url')
        return urlparse.urljoin(base_url, PayzenController._notify_url)

    def _get_languages(self):
        languages = constants.PAYZEN_LANGUAGES
        return [(c, _(l)) for c, l in languages.items()]

    @api.depends('provider')
    def _payzen_compute_multi_warning(self):
        for acquirer in self:
            acquirer.payzen_multi_warning = (constants.PAYZEN_PLUGIN_FEATURES.get('restrictmulti') == True) if (acquirer.provider == 'payzenmulti') else False

    sign_algo_help = _('Algorithm used to compute the payment form signature. Selected algorithm must be the same as one configured in the PayZen Back Office.')

    if constants.PAYZEN_PLUGIN_FEATURES.get('shatwo') == False:
        sign_algo_help += _('The HMAC-SHA-256 algorithm should not be activated if it is not yet available in the PayZen Back Office, the feature will be available soon.')

    # Compatibility with Odoo 14.
    payzen_odoo14 = True if parse_version(release.version) >= parse_version('14') else False

    providers = [('payzen', _('PayZen - Standard payment'))]
    if payzen_odoo14:
        ondelete_policy = {'payzen': 'set default'}

    if constants.PAYZEN_PLUGIN_FEATURES.get('multi') == True:
        providers.append(('payzenmulti', _('PayZen - Payment in installments')))
        if payzen_odoo14:
            ondelete_policy['payzenmulti'] = 'set default'

    if payzen_odoo14:
        provider = fields.Selection(selection_add=providers, ondelete = ondelete_policy)
    else:
        provider = fields.Selection(selection_add=providers)

    payzen_site_id = fields.Char(string=_('Shop ID'), help=_('The identifier provided by PayZen.'), default=constants.PAYZEN_PARAMS.get('SITE_ID'))
    payzen_key_test = fields.Char(string=_('Key in test mode'), help=_('Key provided by PayZen for test mode (available in PayZen Back Office).'), default=constants.PAYZEN_PARAMS.get('KEY_TEST'), readonly=constants.PAYZEN_PLUGIN_FEATURES.get('qualif'))
    payzen_key_prod = fields.Char(string=_('Key in production mode'), help=_('Key provided by PayZen (available in PayZen Back Office after enabling production mode).'), default=constants.PAYZEN_PARAMS.get('KEY_PROD'))
    payzen_sign_algo = fields.Selection(string=_('Signature algorithm'), help=sign_algo_help, selection=[('SHA-1', 'SHA-1'), ('SHA-256', 'HMAC-SHA-256')], default=constants.PAYZEN_PARAMS.get('SIGN_ALGO'))
    payzen_notify_url = fields.Char(string=_('Instant Payment Notification URL'), help=_('URL to copy into your PayZen Back Office > Settings > Notification rules.'), default=_get_notify_url, readonly=True)
    payzen_gateway_url = fields.Char(string=_('Payment page URL'), help=_('Link to the payment page.'), default=constants.PAYZEN_PARAMS.get('GATEWAY_URL'))
    payzen_language = fields.Selection(string=_('Default language'), help=_('Default language on the payment page.'), default=constants.PAYZEN_PARAMS.get('LANGUAGE'), selection=_get_languages)
    payzen_available_languages = fields.Many2many('payzen.language', string=_('Available languages'), column1='code', column2='label', help=_('Languages available on the payment page. If you do not select any, all the supported languages will be available.'))
    payzen_capture_delay = fields.Char(string=_('Capture delay'), help=_('The number of days before the bank capture (adjustable in your PayZen Back Office).'))
    payzen_validation_mode = fields.Selection(string=_('Validation mode'), help=_('If manual is selected, you will have to confirm payments manually in your PayZen Back Office.'), selection=[('-1', _('PayZen Back Office Configuration')), ('0', _('Automatic')), ('1', _('Manual'))])
    payzen_payment_cards = fields.Many2many('payzen.card', string=_('Card types'), column1='code', column2='label', help=_('The card type(s) that can be used for the payment. Select none to use gateway configuration.'))
    payzen_threeds_min_amount = fields.Char(string=_('Disable 3DS'), help=_('Amount below which 3DS will be disabled. Needs subscription to selective 3DS option. For more information, refer to the module documentation.'))
    payzen_redirect_enabled = fields.Selection(string=_('Automatic redirection'), help=_('If enabled, the buyer is automatically redirected to your site at the end of the payment.'), selection=[('0', _('Disabled')), ('1', _('Enabled'))])
    payzen_redirect_success_timeout = fields.Char(string=_('Redirection timeout on success'), help=_('Time in seconds (0-300) before the buyer is automatically redirected to your website after a successful payment.'))
    payzen_redirect_success_message = fields.Char(string=_('Redirection message on success'), help=_('Message displayed on the payment page prior to redirection after a successful payment.'), default=_('Redirection to shop in a few seconds...'))
    payzen_redirect_error_timeout = fields.Char(string=_('Redirection timeout on failure'), help=_('Time in seconds (0-300) before the buyer is automatically redirected to your website after a declined payment.'))
    payzen_redirect_error_message = fields.Char(string=_('Redirection message on failure'), help=_('Message displayed on the payment page prior to redirection after a declined payment.'), default=_('Redirection to shop in a few seconds...'))
    payzen_return_mode = fields.Selection(string=_('Return mode'), help=_('Method that will be used for transmitting the payment result from the payment page to your shop.'), selection=[('GET', 'GET'), ('POST', 'POST')])
    payzen_multi_warning = fields.Boolean(compute='_payzen_compute_multi_warning')

    payzen_multi_count = fields.Char(string=_('Count'), help=_('Total number of payments.'))
    payzen_multi_period = fields.Char(string=_('Period'), help=_('Delay (in days) between payments.'))
    payzen_multi_first = fields.Char(string=_('1st payment'), help=_('Amount of first payment, in percentage of total amount. If empty, all payments will have the same amount.'))

    # Check if it's Odoo 10.
    payzen_odoo10 = True if parse_version(release.version) < parse_version('11') else False

    # Compatibility betwen Odoo 13 and previous versions.
    payzen_odoo13 = True if parse_version(release.version) >= parse_version('13') else False

    if payzen_odoo13:
        image = fields.Char()
        environment = fields.Char()
    else:
        image_128 = fields.Char()
        state = fields.Char()

    payzen_redirect = False

    @api.model
    def multi_add(self, filename):
        file = path.join(path.dirname(path.dirname(path.abspath(__file__)))) + filename

        if (constants.PAYZEN_PLUGIN_FEATURES.get('multi') == True):
            convert_xml_import(self._cr, 'payment_payzen', file)

        return None

    def _get_ctx_mode(self):
        ctx_key = self.state if self.payzen_odoo13 else self.environment
        ctx_value = 'TEST' if ctx_key == 'test' else 'PRODUCTION'

        return ctx_value

    def _payzen_generate_sign(self, acquirer, values):
        key = self.payzen_key_prod if self._get_ctx_mode() == 'PRODUCTION' else self.payzen_key_test

        sign = ''
        for k in sorted(values.keys()):
            if k.startswith('vads_'):
                sign += values[k] + '+'

        sign += key

        if self.payzen_sign_algo == 'SHA-1':
            shasign = sha1(sign.encode('utf-8')).hexdigest()
        else:
            shasign = base64.b64encode(hmac.new(key.encode('utf-8'), sign.encode('utf-8'), sha256).digest()).decode('utf-8')

        return shasign

    def _get_payment_config(self, amount):
        if self.provider == 'payzenmulti':
            if (self.payzen_multi_first):
                first = int(float(self.payzen_multi_first) / 100 * int(amount))
            else:
                first = int(float(amount) / float(self.payzen_multi_count))

            payment_config = u'MULTI:first=' + str(first) + u';count=' + self.payzen_multi_count + u';period=' + self.payzen_multi_period
        else:
            payment_config = u'SINGLE'

        return payment_config

    def payzen_form_generate_values(self, values):
        base_url = self.env['ir.config_parameter'].get_param('web.base.url')

        # trans_id is the number of 1/10 seconds from midnight.
        now = datetime.now()
        midnight = now.replace(hour = 0, minute = 0, second = 0, microsecond = 0)
        delta = int((now - midnight).total_seconds() * 10)
        trans_id = str(delta).rjust(6, '0')

        threeds_mpi = u''
        if self.payzen_threeds_min_amount and float(self.payzen_threeds_min_amount) > values['amount']:
            threeds_mpi = u'2'

        # Check currency.
        currency_num = tools.find_currency(values['currency'].name)
        if currency_num is None:
            _logger.error('The plugin cannot find a numeric code for the current shop currency {}.'.format(values['currency'].name))
            raise ValidationError(_('The shop currency {} is not supported.').format(values['currency'].name))

        # Amount in cents.
        k = int(values['currency'].decimal_places)
        amount = int(float_round(float_round(values['amount'], k) * (10 ** k), 0))

        # List of available languages.
        available_languages = ''
        for value in self.payzen_available_languages:
            available_languages += value.code + ';'

        # List of available payment cards.
        payment_cards = ''
        for value in self.payzen_payment_cards:
            payment_cards += value.code + ';'

        #Validation mode
        validation_mode = self.payzen_validation_mode if self.payzen_validation_mode != '-1' else ''

        # Enable redirection?
        AcquirerPayzen.payzen_redirect = True if str(self.payzen_redirect_enabled) == '1' else False

        tx_values = dict() # Values to sign in unicode.
        tx_values.update({
            'vads_site_id': self.payzen_site_id,
            'vads_amount': str(amount),
            'vads_currency': currency_num,
            'vads_trans_date': str(datetime.utcnow().strftime("%Y%m%d%H%M%S")),
            'vads_trans_id': str(trans_id),
            'vads_ctx_mode': str(self._get_ctx_mode()),
            'vads_page_action': u'PAYMENT',
            'vads_action_mode': u'INTERACTIVE',
            'vads_payment_config': self._get_payment_config(amount),
            'vads_version': constants.PAYZEN_PARAMS.get('GATEWAY_VERSION'),
            'vads_url_return': urlparse.urljoin(base_url, PayzenController._return_url),
            'vads_order_id': str(values.get('reference')),
            'vads_contrib': constants.PAYZEN_PARAMS.get('CMS_IDENTIFIER') + u'_' + constants.PAYZEN_PARAMS.get('PLUGIN_VERSION') + u'/' + release.version,

            'vads_language': self.payzen_language or '',
            'vads_available_languages': available_languages,
            'vads_capture_delay': self.payzen_capture_delay or '',
            'vads_validation_mode': validation_mode,
            'vads_payment_cards': payment_cards,
            'vads_return_mode': str(self.payzen_return_mode),
            'vads_threeds_mpi': threeds_mpi,

            # Customer info.
            'vads_cust_id': str(values.get('billing_partner_id')) or '',
            'vads_cust_first_name': values.get('billing_partner_first_name') and values.get('billing_partner_first_name')[0:62] or '',
            'vads_cust_last_name': values.get('billing_partner_last_name') and values.get('billing_partner_last_name')[0:62] or '',
            'vads_cust_address': values.get('billing_partner_address') and values.get('billing_partner_address')[0:254] or '',
            'vads_cust_zip': values.get('billing_partner_zip') and values.get('billing_partner_zip')[0:62] or '',
            'vads_cust_city': values.get('billing_partner_city') and values.get('billing_partner_city')[0:62] or '',
            'vads_cust_state': values.get('billing_partner_state').code and values.get('billing_partner_state').code[0:62] or '',
            'vads_cust_country': values.get('billing_partner_country').code and values.get('billing_partner_country').code.upper() or '',
            'vads_cust_email': values.get('billing_partner_email') and values.get('billing_partner_email')[0:126] or '',
            'vads_cust_phone': values.get('billing_partner_phone') and values.get('billing_partner_phone')[0:31] or '',

            # Shipping info.
            'vads_ship_to_first_name': values.get('partner_first_name') and values.get('partner_first_name')[0:62] or '',
            'vads_ship_to_last_name': values.get('partner_last_name') and values.get('partner_last_name')[0:62] or '',
            'vads_ship_to_street': values.get('partner_address') and values.get('partner_address')[0:254] or '',
            'vads_ship_to_zip': values.get('partner_zip') and values.get('partner_zip')[0:62] or '',
            'vads_ship_to_city': values.get('partner_city') and values.get('partner_city')[0:62] or '',
            'vads_ship_to_state': values.get('partner_state').code and values.get('partner_state').code[0:62] or '',
            'vads_ship_to_country': values.get('partner_country').code and values.get('partner_country').code.upper() or '',
            'vads_ship_to_phone_num': values.get('partner_phone') and values.get('partner_phone')[0:31] or '',
        })

        if AcquirerPayzen.payzen_redirect:
            tx_values.update({
                'vads_redirect_success_timeout': self.payzen_redirect_success_timeout or '',
                'vads_redirect_success_message': self.payzen_redirect_success_message or '',
                'vads_redirect_error_timeout': self.payzen_redirect_error_timeout or '',
                'vads_redirect_error_message': self.payzen_redirect_error_message or ''
            })

        payzen_tx_values = dict() # Values encoded in UTF-8.

        for key in tx_values.keys():
            if tx_values[key] == ' ':
                tx_values[key] = ''

            payzen_tx_values[key] = tx_values[key].encode('utf-8')

        payzen_tx_values['payzen_signature'] = self._payzen_generate_sign(self, tx_values)
        return payzen_tx_values

    def payzenmulti_form_generate_values(self, values):
        return self.payzen_form_generate_values(values)

    def payzen_get_form_action_url(self):
        return self.payzen_gateway_url

    def payzenmulti_get_form_action_url(self):
        return self.payzen_gateway_url

class TransactionPayzen(models.Model):
    _inherit = 'payment.transaction'

    payzen_trans_status = fields.Char(_('Transaction status'))
    payzen_card_brand = fields.Char(_('Means of payment'))
    payzen_card_number = fields.Char(_('Card number'))
    payzen_expiration_date = fields.Char(_('Expiration date'))
    payzen_auth_result = fields.Char(_('Authorization result'))
    payzen_raw_data = fields.Text(string=_('Transaction log'), readonly=True)

    # --------------------------------------------------
    # FORM RELATED METHODS
    # --------------------------------------------------

    @api.model
    def _payzen_form_get_tx_from_data(self, data):
        shasign, status, reference = data.get('signature'), data.get('vads_trans_status'), data.get('vads_order_id')

        if not reference or not shasign or not status:
            error_msg = 'PayZen : received bad data {}'.format(data)
            _logger.error(error_msg)
            raise ValidationError(error_msg)

        tx = self.search([('reference', '=', reference)])
        if not tx or len(tx) > 1:
            error_msg = 'PayZen: received data for reference {}'.format(reference)
            if not tx:
                error_msg += '; no order found'
            else:
                error_msg += '; multiple order found'

            _logger.error(error_msg)
            raise ValidationError(error_msg)

        # Verify shasign.
        shasign_check = tx.acquirer_id._payzen_generate_sign('out', data)
        if shasign_check.upper() != shasign.upper():
            error_msg = 'PayZen: invalid shasign, received {}, computed {}, for data {}'.format(shasign, shasign_check, data)
            _logger.info(error_msg)
            raise ValidationError(error_msg)

        return tx

    def _payzen_form_get_invalid_parameters(self, data):
        invalid_parameters = []

        # Check what is bought.
        amount = float(int(data.get('vads_amount', 0)) / math.pow(10, int(self.currency_id.decimal_places)))

        if float_compare(amount, self.amount, int(self.currency_id.decimal_places)) != 0:
            invalid_parameters.append(('amount', amount, '{:.2f}'.format(self.amount)))

        currency_code = tools.find_currency(self.currency_id.name)
        if (currency_code is None) or (int(data.get('vads_currency')) != int(currency_code)):
            invalid_parameters.append(('currency', data.get('vads_currency'), currency_code))

        return invalid_parameters

    def _payzen_form_validate(self, data):
        payzen_statuses = {
            'success': ['AUTHORISED', 'CAPTURED', 'ACCEPTED'],
            'pending': ['AUTHORISED_TO_VALIDATE', 'WAITING_AUTHORISATION', 'WAITING_AUTHORISATION_TO_VALIDATE', 'INITIAL', 'UNDER_VERIFICATION', 'WAITING_FOR_PAYMENT', 'PRE_AUTHORISED'],
            'cancel': ['ABANDONED']
        }

        html_3ds = _('3DS authentication: ')
        if data.get('vads_threeds_status') == 'Y':
            html_3ds += _('YES')
            html_3ds += '<br />' + _('3DS certificate: ') + data.get('vads_threeds_cavv')
        else:
            html_3ds += _('NO')

        expiry = ''
        if data.get('vads_expiry_month') and data.get('vads_expiry_year'):
            expiry = data.get('vads_expiry_month').zfill(2) + '/' + data.get('vads_expiry_year')

        values = {
            'acquirer_reference': data.get('vads_trans_uuid'),
            'payzen_raw_data': '{}'.format(data),
            'html_3ds': html_3ds,
            'payzen_trans_status': data.get('vads_trans_status'),
            'payzen_card_brand': data.get('vads_card_brand'),
            'payzen_card_number': data.get('vads_card_number'),
            'payzen_expiration_date': expiry,
        }

        # Set validation date.
        key = 'date' if hasattr(self, 'date')  else 'date_validate'
        values[key] = fields.Datetime.now()

        status = data.get('vads_trans_status')
        if status in payzen_statuses['success']:
            values.update({
                'state': 'done',
            })

            self.write(values)

            return True
        elif status in payzen_statuses['pending']:
            values.update({
                'state': 'pending',
            })

            self.write(values)

            return True
        elif status in payzen_statuses['cancel']:
            self.write({
                'state_message': 'Payment for transaction #%s is cancelled (%s).' % (self.reference, data.get('vads_result')),
                'state': 'cancel',
            })

            return False
        else:
            auth_result = data.get('vads_auth_result')
            auth_message = _('See the transaction details for more information ({}).').format(auth_result)

            error_msg = 'PayZen payment error, transaction status: {}, authorization result: {}.'.format(status, auth_result)
            _logger.info(error_msg)

            values.update({
                'state_message': 'Payment for transaction #%s is refused (%s).' % (self.reference, data.get('vads_result')),
                'state': 'error',
                'payzen_auth_result': auth_message,
            })

            self.write(values)

            return False
